### ISSUE TITLE
Bug: no-useless-default-assignment false positive when implementation parameter type doesn't include undefined but outer variable's conditional type makes it optional

### ISSUE THREAD

**Issue by lyhoang-web-dev** [typescript-eslint/typescript-eslint#12078]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=C4TwDgpgBAYhwGMAWMB2AeAKgBQIYCdcBbAZwD4oBeAWACgoo99gBLXAGyyePKggA9gEVABMSUHAR50GAfigAKMFNKyAXBO6kAlFQoA3APYsRMqBqUqSGyYR16oRkwG46dAGYBXVAlaHUUAj4ELhCcIhIEPhcVmQKugDeZqCQjFZUmlautGYA9LlQAKL4%2BIb4UO4sweKRwRoARAAiEO64nuzAjhye0CziniQQ7BAk4gBGEAhtg1DAkVDKdvBRUH1QqIadhmB%2BqBz1eQXhyGgxdrxruFAiLVHBIoH%2BIiy7HLPg0IAoBEUAygAyLFQW1Q7BAUEGI1m8wEYHYLAQL0OUFwqA2wFCLH8UAABkxWBx0FpyNjFBscd4bpVUBARNjdCiHoCEGVgr5QRV2LgAObiObQSltDoAOiRACFPJ0AO6RAK2HhQJC4cQcdgAWm2r3YFRYQzEABpYPAThg5aQKJRFIseOpMuddJQDMYRHqkUQQqhAVzAiqouIpgFDEQXgt0gyofyWoLOgBJH5QYK4ZC4MbDEX0R6oEiddyoDTHFAmonmy1WDR4ticIsZBIAX2R4nLBKL9ooSXTDGZmcMqfYhi5ljt2QYNeyZmCwE8%2BACOeyNboQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tmUSWWS0AJogBmAQwR1xAygHMmAW0Qt0URNGgdokcGAC%2BIfUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEUJemy0eFY9AruuGAL4jdA&tokens=false

### Repro Code

```TypeScript
type FetchFn<TParams> =
  Partial<TParams> extends TParams
    ? (params?: TParams) => void
    : (params: TParams) => void;

function createFetcher<TParams>() {
  type Params = TParams;

  // Error fires here: "Default value is useless because the parameter is not optional"
  // FetchFn<TParams> is a deferred conditional type — ESLint only sees the explicit
  // annotation `Partial<Params>` (no `undefined`) and incorrectly flags the default.
  // But when TParams has all-optional fields, FetchFn<TParams> = (params?: TParams) => void,
  // meaning callers can omit params and the default IS reachable.
  const fn: FetchFn<TParams> = (params: Partial<Params> = {} as Partial<Params>) => {
    console.log(params);
  };

  return fn;
}
```

### ESLint Config

```javascript
{
"rules": {
    "@typescript-eslint/no-useless-default-assignment": "error"
 }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No ESLint error. `FetchFn<TParams>` is a conditional type — when `Partial<TParams> extends TParams` (all fields optional), it resolves to `(params?: TParams) => void`. The default `= {}` is reachable and not useless.

### Actual Result

Default value is useless because the parameter is not optional.
`@typescript-eslint/no-useless-default-assignment`

### Additional Info

The false positive occurs because FetchFn<TParams> is a deferred conditional type (generic `TParams` is unresolved at the implementation site). ESLint only inspects the explicit parameter annotation `Partial<Params>` — which has no `undefined ` — and incorrectly concludes the default is useless.

With a concrete type the rule behaves correctly:
```tsx
type AllOptional = { id?: string };
// FetchFn<AllOptional> resolves immediately → (params?: AllOptional) → no error ✓
const fn: FetchFn<AllOptional> = (params: AllOptional = {}) => {};
```
The false positive only triggers when the conditional type is deferred due to unresolved generic parameters.

- `TypeScript`: 5.9.3
- `ESLint`: 9.39.2
- `typescript-eslint`: 8.56.0

**Comment by JoshuaKGoldberg** [comment]
What a fun tricky reproduction! +1 that this is a bug. Accepting PRs.

**Comment by lyhoang-web-dev** [comment]
@JoshuaKGoldberg version 8.57.1 is still an issue. Could you check the Playground link to see that?
